docs/readmes/proposals/p024_magma_settlement_service.md

```diff
@@ -0,0 +1,281 @@
+
+- Step [1-2] The State channel opening procedure. The Helium Handler sends Open State Channel Request to open the State Channel. State Channel is opened once the positive Response is received from State Channel.
+- Step [3] On this Step the Handler sends the Used Data Credits for each Used Data Quota per Miner pub key.
+- Step [4-5] The State channel closure procedure. The closure procedure contains two type of triggers - all the Data credits have been used from the State channel and end of an Epoch.
+
+#### 3.2.2 UE attach procedure
+
+Following diagram shows e2e UE attach scenario that will be in real life. It consists of existing functionality (marked with black), new functionality to be implemented in magma (marked with blue).
+The attach procedure implies that State Channel is already in open state.
+
+![Attach call flow](https://user-images.githubusercontent.com/93994458/149513666-63ae0f4b-c8be-4cfc-a739-909723fca3aa.png)
+
+This scenario consists of the steps below:
+
+- Step [1 - 2] Initial attach procedure in a mobile network. The UE establishes radio link synchronization with an eNB, the UE creates a connection for data delivery via an Attach Request message sent to the eNB, the eNB then forwards the attach request to an MME/SGW.
+- Step [3] The SGW sends CCR-I to the Session Purchaser requesting volume quota.
+- Step [4] Session Purchaser checks if the IMSI-prefix (i.e. the HPLMN operator) has balance to answer with granted data units on CCR-I Request which has been received on Step [3].
+- Step [5] In response to a CCR-I (Step [3]), the Session Purchaser returns a CCA-I message that indicates success (DIAMETER_SUCCESS) or failure (DIAMETER AUTHORIZATION REJECTED) depending on whether the HPLMN operator has sufficient credit for the requested services. In case of HPLMN operator has sufficient credit the CCA contains the Granted Service Unit in Bytes.
+- Step [6] The SGW sends signed Session Init Event to Miner. The message is signed with SGW PubKey.
+- Step [7] The Miner sends signed Session Init Event to Helium Handler. The message is signed with Miner PubKey.
+- Step [8 - 9] The Create Session Request and Response messages to establish UE requested PDN connectivity through the SGW and Home PGW.
+- Step [10-15] Establishing radio link between UE and eNodeB for packet data transfer.
+- Step [16-17] Packet data between UE and Internet.
+
+#### 3.2.3 UE quota update procedure
+
+Following diagram shows e2e UE update scenario that will be in real life. It consists of existing functionality (marked with black), new functionality to be implemented in magma and helium (marked with blue).
+
+![Update call flow](https://user-images.githubusercontent.com/93994458/149513805-f78e1c18-bb12-4f1f-abf9-0a0ea5e09d20.png)
+
+This scenario consists of the steps below:
+
+- Step [1-2] Packet data transfer is ongoing between UE and Internet.
+- Step [3] The threshold of the quota is configured on SGW (e.g. 95% of granted service unit).
```

Comment:
Currently on inbound roaming we do not use sessiond. We create a flow directly and it is the other operators PGW who does keep track of the data.

To achieve that you could try to refactor inbound roaming to use sessiond. Then sessiond create an appropriate flow for that UE with a hardcoded pass all rule tracking for all the traffic.

That would requiere changes on mme, sessiond and pipelined (to use the proper API call for roaming subscribers)

Reply:
sessiond would then send the updates to session_proxy or to your own implementation of session_proxy (session_purchaser ?)

Reply:
that was our initial approach to  refactor sessiond to use it to push data to session_proxy.
If you may advise us better approach - you are welcome.